[Collections-834][bloom filters] bit count producer operation is not clearly defined

Claudenw opened the pull request:
fix for COLLECTIONS-384

Document bit count producer operation in the BitCountProducer.java file.

Add unit tests for all BitCountProducers.

Claudenw committed:
Added tests for BitCountProducer

Claudenw committed:
fixed checkstyle issues

Claudenw committed:
updated javadoc

Claudenw committed:
cleaned up formatting

Claudenw committed:
removed extraneous files

Claudenw commented:
@aherbert Described the BitCountProducer contract in the BitCountProducer javadoc, and added tests to show that they work as expected.  I hope I have not made too much of a hash of it.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/BitCountProducerFromDefaultIndexProducerTest.java:
```diff
@@ -41,24 +37,19 @@ protected int getBehaviour() {
-        });
+    @Override
+    protected int[] getExpectedIndex() {
+        return expected;
```
If the Abstract test is modified as suggested to separately test behaviour and the expected indices then this can `return data`.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/BitCountProducerFromIntArrayTest.java:
```diff
@@ -0,0 +1,54 @@
+public class BitCountProducerFromIntArrayTest extends AbstractBitCountProducerTest {
+
+    int[] data = {6, 8, 1, 2, 4, 4, 5};
+    int[] expected = {1, 2, 4, 5, 6, 8};
```
Updating the abstract test removes the requirement for `expected`

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/BitCountProducerFromHasherTest.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    protected BitCountProducer createProducer() {
+        return BitCountProducer.from(new IncrementingHasher(0, 1).indices(Shape.fromKM(17, 72)));
```
I would update the test to use something other than IncrementingHasher(0, 1), perhaps IncrementingHasher(3, 1), or IncrementingHasher(3, 2). This would check a non-sequential set of indices.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/BitCountProducerFromSparseBloomFilterTest.java:
```diff
@@ -0,0 +1,55 @@
+        BloomFilter bf = new SparseBloomFilter(shape);
+        bf.merge(hasher);
+        return BitCountProducer.from(bf);
+
```
Remove empty line

Claudenw replied:
fixed

aherbert reviewed (changes requested):
Hi @Claudenw,

Thanks for all the new test data. I had failed to spot previously that the expected indices from an IndexProducer are not actually checked. I think these changes address the issue with the disabled test in BitCountProducerFromIndexProducerTest and strengthen the test suite.

You test the expected indices using an array that mandates the order. However we already test the order behaviour separately. I suggested a change to only use the expected indices to detect correct/incorrect actual indices. The remaining behaviour tests will detect ordering and duplication problems.

I provided an implementation for AbstractBitCountProducerTest that matches the AbstractIndexProducerTest. This checks consistency between the indices output from the two interface forEach methods. Currently it includes new behaviour flags for forEachCount. However the test suite passes on the assumption the forEachIndex and forEachCount have the same behaviour. So maybe we can drop them. They could be added in the future if an implementation is created with different behaviour for the two methods.

I would change some of the tests from using IncrementingHasher(0, 1) to something else. This avoids a sequence from 0 which is simplistic test data. The data is easy to create with a stream:
```Java
    // For: IncrementingHasher(3, 2) and Shape.fromKM(17, 72)
    IntStream.iterate(3, i -> i + 2).limit(17).toArray();
```
For multiple hashers you can collate them:
```Java
    // For: IncrementingHasher(3, 2), IncrementingHasher(0, 1), and Shape.fromKM(17, 72)
    IntStream.concat(
        IntStream.iterate(3, i -> i + 2).limit(17),
        IntStream.iterate(0, i -> i + 1).limit(17)
    ).distinct().toArray();
```

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/BitCountProducer.java:
```diff
@@ -47,7 +60,7 @@ default boolean forEachIndex(IntPredicate predicate) {
     /**
      * Creates a BitCountProducer from an IndexProducer.  The resulting
-     * producer will count each enabled bit once.
+     * producer will return every index from the IndexProducer with a count of 1.
```
added with reference to class javaodc

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -93,6 +92,12 @@ int[] toArray() {
      */
     protected abstract int getBehaviour();
 
+    protected abstract int[] getExpectedIndex();
```
added

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -93,6 +92,12 @@ int[] toArray() {
 
+    protected abstract int[] getExpectedIndex();
+
+    protected int[] getExpectedForEach() {
```
removed

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -93,6 +92,12 @@ int[] toArray() {
+    }
+
     @Test
     public final void testForEachIndex() {
```
renamed

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -144,12 +148,13 @@ public final void testBehaviourAsIndexArray() {
             long count = Arrays.stream(actual).distinct().count();
             Assertions.assertEquals(count, actual.length);
         }
+        int[] expected = getExpectedIndex();
```
moved to testAsIndexArray

Claudenw replied:
removed

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -161,6 +166,8 @@ public final void testBehaviourForEach() {
             long count = Arrays.stream(actual).distinct().count();
             Assertions.assertEquals(count, actual.length);
         }
+        int[] expected = getExpectedForEach();
```
Moved to its own testForEachIndex

Claudenw committed:
Updated as per requests

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -149,7 +181,6 @@ public final void testBehaviourAsIndexArray() {
     @Test
     public final void testBehaviourForEach() {
         int flags = getBehaviour();
-        Assumptions.assumeTrue((flags & (FOR_EACH_ORDERED | FOR_EACH_DISTINCT)) != 0);
```
Reinstate the Assumptions

Claudenw replied:
fixed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractBitCountProducerTest.java:
```diff
@@ -16,35 +16,29 @@
-            return true;
-        }
-    };
+    protected abstract int[][] getExpectedBitCount();
```
This method may not need to be abstract. In most cases where ordering and duplicates are not mandated then this implementation will be fine:
```Java
    protected int[][] getExpectedBitCount() {
        return Arrays.stream(getExpectedIndices()).mapToObj(x -> new int[] {x, 1}).toArray(int[][]::new);
    }
```

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -93,8 +98,36 @@ int[] toArray() {
+     * Creates an array of expected indices.
+     * @return an array of expected indices.
+     */
+    protected abstract int[] getExpectedIndices();
```
Note that this method requires assumptions. The expected indices are dependent on the producer returned from createProducer, which is ultimately dependent on the shape of the Bloom filter.

Claudenw replied:
fixed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/IndexProducerFromIntArrayTest.java:
```diff
@@ -18,19 +18,32 @@
+        return expected;
+    }
+
+//    @Override
```
Remove commented out code

Claudenw replied:
fixed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/BitCountProducerFromIntArrayTest.java:
```diff
@@ -0,0 +1,49 @@
+public class BitCountProducerFromIntArrayTest extends AbstractBitCountProducerTest {
+
+    int[] data = {6, 8, 1, 2, 4, 4, 5};
+
```
Remove extra line

Claudenw replied:
fixed

aherbert reviewed (changes requested):
I suggested some changes to the AbstractBitCountProducerTest to assert on the behaviour flags separately from the bit count totals. Can you add these? Please see [this note](https://github.com/apache/commons-collections/pull/335#discussion_r972942396).

Claudenw committed:
Made recommended corrections

Claudenw committed:
fixed checkstyle issues

Claudenw committed:
added documentation

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -93,8 +98,36 @@ int[] toArray() {
+
+    @Test
+    public final void testAsIndexArrayValues() {
+        List<Integer> lst = new ArrayList<>();
```
fixed

aherbert replied:
You have not removed the redundant part of this test. The distinctness is tested in `testBehaviourAsIndexArray`.

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -43,7 +48,7 @@ public abstract class AbstractIndexProducerTest {
      * An expandable list of int values.
      */
-    private static class IntList {
+    protected static class IntList {
```
Must be protected for AbstractBitCountProducer usage

aherbert replied:
When I checked you had not pushed the entire AbstractBitCountProducerTest and so it was not (yet) used in that class. I see the changes now. Let me know when you have fixed everything and I'll review again.

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractBitCountProducerTest.java:
```diff
@@ -16,35 +16,29 @@
 import org.apache.commons.collections4.bloomfilter.BitCountProducer.BitCountConsumer;
 import org.junit.jupiter.api.Test;
 
 public abstract class AbstractBitCountProducerTest extends AbstractIndexProducerTest {
```
I don't know what happened.  I have, I think, fixed it now.

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -134,7 +167,6 @@ public final void testConsistency() {
     @Test
     public final void testBehaviourAsIndexArray() {
         int flags = getBehaviour();
-        Assumptions.assumeTrue((flags & (AS_ARRAY_ORDERED | AS_ARRAY_DISTINCT)) != 0);
```
fixed

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -93,8 +98,36 @@ int[] toArray() {
+
     @Test
     public final void testForEachIndex() {
+        //IndexProducer producer = createProducer();
```
fixed

Claudenw committed:
fixed line end issues

aherbert reviewed (changes requested):
Hi @Claudenw 

I've not looked at everything here. I found an issue where the update to the documentation of `BitCountProducer` is not correct (given the current implementations). See my test that is checking the count of indices output from the IndexProducer. It fails for `ArrayCountingBloomFilter`.

I don't think we want to change the implementations. We should change the documentation to be less strict. You can then remove some of the updates to the tests as we do not need `int[] getExpectedForEach`. It is only overridden in `BitCountProducerFromHasherTest` anyway.

Since you have not changed any of `src/main` besides javadoc I can only assume this was to pass the test that I think is not required if we are less strict with the interface definition.

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -134,8 +176,8 @@ public final void testConsistency() {
         int flags = getBehaviour();
-        Assumptions.assumeTrue((flags & (AS_ARRAY_ORDERED | AS_ARRAY_DISTINCT)) != 0);
         int[] actual = createProducer().asIndexArray();
+        assumeTrue((flags & (AS_ARRAY_ORDERED | AS_ARRAY_DISTINCT)) != 0);
```
fixed

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractIndexProducerTest.java:
```diff
@@ -149,7 +191,6 @@ public final void testBehaviourAsIndexArray() {
     @Test
     public final void testBehaviourForEach() {
         int flags = getBehaviour();
-        Assumptions.assumeTrue((flags & (FOR_EACH_ORDERED | FOR_EACH_DISTINCT)) != 0);
```
fixed

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractBitCountProducerTest.java:
```diff
@@ -60,23 +80,90 @@ public boolean test(int index, int count) {
+    @Test
+    public final void testBehaviourForEachCount() {
+        int flags = getBehaviour();
+        IntList list = new IntList();
```
fixed

aherbert replied:
As before, assumeTrue should be in the line after you get the flags.

Claudenw commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractBitCountProducerTest.java:
```diff
@@ -60,23 +80,90 @@ public boolean test(int index, int count) {
+            long count = Arrays.stream(actual).distinct().count();
+            Assertions.assertEquals(count, actual.length);
         }
+        int[] expected = getExpectedForEach();
```
At first I thought you were wrong, but now I see we are in violent agreement and my documentation is poorly written! 

Given the example at the bottom of the BitCountProducer interface documentation.

```
 * An IndexProducer that generates the values [1,2,3,1,5,6] can be represented by a BitCountProducer
 * that generates [(1,2),(2,1),(3,1),(5,1)(6,1)] or [(1,1),(2,1),(3,1),(1,1),(5,1),(6,1)] where the
 * entries may be in any order.
```

How would you rephrase:

```
 * A BitCountProducer may return duplicate indices and may be unordered.
 *
 * The guarantees are:
 * <ul>
 * <li>that for every unique value produced by the IndexProducer there will be at least one
 * index in the BitCountProducer.</li>
 * <li>that the total count of a specific value produced by the IndexProducer will equal the
 * total of the counts in the BitCountProducer for that index.</li>
 * </ul>
```

I have removed the `getExpectedForEach()` from the tests.

aherbert replied:
I think this is what we have:

```
 * Defines a mapping of index to counts.
 *
 * <p>Note that a BitCountProducer may return duplicate indices and may be unordered.
 *
 * <p>Implementations must guarantee that:
 *
 * <ul>
 * <li>The mapping of index to counts is the combined sum of counts at each index.
 * <li>For every unique value produced by the IndexProducer there will be at least one matching
 * index and count produced by the BitCountProducer.
 * <li>The BitCountProducer will not generate indices that are not output by the IndexProducer.
 * </ul>
 *
 * <p>Note that implementations that do not output duplicate indices for BitCountProducer and
 * do for IndexProducer, or vice versa, are consistent if the distinct indices from each are
 * the same.
 *
 * <p>For example the mapping [(1,2),(2,3),(3,1)] can be output with many combinations including:
 * <pre>
 * [(1,2),(2,3),(3,1)]
 * [(1,1),(1,1),(2,1),(2,1),(2,1),(3,1)]
 * [(1,1),(3,1),(1,1),(2,1),(2,1),(2,1)]
 * [(3,1),(1,1),(2,2),(1,1),(2,1)]
 * ...
 * </pre>
```

I am not sure we need to spell it out with the example. In this case perhaps less is more.

Claudenw replied:
Replaced javadoc text with this code..

Claudenw committed:
Removed getExpectedForEach() added missing AssumeTrue() checks

Claudenw committed:
Updated javadoc

Claudenw committed:
Merge branch 'apache:master' into Collections-384_BitCountProducer_operation_is_not_clearly_defined

aherbert commented on src/main/java/org/apache/commons/collections4/bloomfilter/BitCountProducer.java:
```diff
@@ -47,7 +72,7 @@ default boolean forEachIndex(IntPredicate predicate) {
     /**
      * Creates a BitCountProducer from an IndexProducer.  The resulting
-     * producer will count each enabled bit once.
+     * producer will return every index from the IndexProducer with a count of 1.
```
I was considering why do we need to do this: convert an IndexProducer to a BitCountProducer. It would be to add it to a CountingBloomFilter. In this case the conversion does not remove duplicate indices. This means that the BitCountProducer can increment the count of an index more than 1 when adding a single IndexProducer.

This has the following consequences:
- The count at each index of a CountingBloomFilter does not represent the number of individual hashes added to the filter that had that index enabled. The count is bounded by the number of hashes as a lower limit, i.e. it may be higher.
- Removal of an IndexProducer from a counting bloom filter must use the same method to construct the BitCountProducer as the addition in order to use the exact same index->count mapping.

I am fine with this behaviour. Filtering an IndexProducer to unique indices is extra work. The `IndexFilter` class helps with this but requires a `Shape`. So it cannot be done in the BitCountProducer interface which has no Shape; other mechanisms to filter the IndexProducer without a bound on the index would be inefficient.

So perhaps we require a bit of extra documentation here to serve as a user-beware warning:
```
 * <p>Note that the BitCountProducer does not remove duplicates. Any use of the
 * BitCountProducer to create an aggregate mapping of index to counts, such as a
 * CountingBloomFilter, should use the same BitCountProducer in both add and
 * subtract operations to maintain consistency.
```
I believe this is consistent with the new wording in the javadoc header for the BitCountProducer interface.

Claudenw replied:
added

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractBitCountProducerTest.java:
```diff
@@ -16,35 +16,46 @@
 public abstract class AbstractBitCountProducerTest extends AbstractIndexProducerTest {
 
+    /** Flag to indicate the {@link BitCountProducer#forEachCount(BitCountConsumer)} is ordered. */
+    protected static final int FOR_EACH_COUNT_ORDERED = 0x10;
```
In the interest of simplicity, can these be dropped?

I previously stated that we should only have them if the behaviour of a BitCountProducer will be different from an IndexProducer. From what I can see in the tests these are used in 3 tests, and only when `FOR_EACH_ORDERED` and `FOR_EACH_DISTINCT` from the AbstractIndexProducerTest are also used. Dropping these constants will reduce maintenance and ensure the flags are correctly set for BitCount and Index producers.

You can add a note in the test to state that the same flags are reused.

I think the simplest way to do this is to update the flags to be private:
```Java
/** Flag to indicate the {@link BitCountProducer#forEachCount(BitCountConsumer)} is ordered.
 * This flag currently reuses the value from IndexProducer. At present no implementations
 * exhibit different behaviour for the two interfaces. This may change in the future and
 * an explicit flag value will be required. */
private static final int FOR_EACH_COUNT_ORDERED = FOR_EACH_ORDERED;
/** Flag to indicate the {@link BitCountProducer#forEachCount(BitCountConsumer)} is distinct.
 * This flag currently reuses the value from IndexProducer. At present no implementations
 * exhibit different behaviour for the two interfaces. This may change in the future and
 * an explicit flag value will be required. */
private static final int FOR_EACH_COUNT_DISTINCT = FOR_EACH_DISTINCT;
```
An alternative would be to have a flag for unspecified behaviour. This must be explicitly set by implementing tests:
```
FOR_EACH_COUNT_UNSPECIFIED = 0x10;
FOR_EACH_COUNT_ORDERED = 0x20;
FOR_EACH_COUNT_DISTINCT = 0x40;
```
The abstract test can then check if the behaviour is unspecified. If true then the other flags must be unset. If false then at least one of the other flags must be set. I think this is more work and not required, especially since we are testing the behaviour that is not a part of the contract given in the javadoc of the main implementations. This is more of a behaviour sense check of the implementations.

Claudenw replied:
I'm not sure I like this, but I can't think of a good counter argument.  I could do the hand waving and say IndexProducer and BitCountProducer produce different orders but I can't see a reasonable example in any possible implementation I can come up with.  So, changes made.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/AbstractBitCountProducerTest.java:
```diff
@@ -60,23 +71,89 @@ public boolean test(int index, int count) {
-            assertTrue(createEmptyProducer().forEachCount(FALSE_CONSUMER), "empty should be true");
-            assertTrue(createEmptyProducer().forEachCount(TRUE_CONSUMER), "empty should be true");
+    @Test
+    public final void testBehaviourForEachCount() {
```
Add a javadoc note:
```Java
/**
 * Test the behaviour of {@link BitCountProducer#forEachCount(BitCountConsumer)} with respect
 * to ordered and distinct indices. Currently the behaviour is assumed to be the same as
 * {@link IndexProducer#forEachIndex(java.util.function.IntPredicate)}.
 */
```

Claudenw replied:
added

aherbert reviewed (commented):
Hi @Claudenw 

This is looking good. I suggested a bit more text for the BitCountProducer interface. There is also a simplification we can make in the AbstractBitCountProducerTest to lower maintenance by hiding the new flag constants (for now).

Claudenw committed:
Cleaned up as per requests.

Claudenw committed:
Added tests for hasher unique and absoluetUnique IndexProducer and
BitCountProducer.

Claudenw committed:
fixed formatting

Claudenw committed:
fixed checkstyle issues

Claudenw commented:
I discovered issues with the HasherCollection and Hasher generated IndexProducer and BitCountProducers created from those IndexProducers.  Additional tests and corrections included in this latest push.

aherbert commented:
Last commit was on Oct 21st. Is this correct or do you need to push again?

Claudenw commented:
Let me see what happened.  In pushed this AM.  (or so I thought)


On Mon, Oct 24, 2022, 11:57 Alex Herbert ***@***.***> wrote:

> Last commit was on Oct 21st. Is this correct or do you need to push again?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-collections/pull/335#issuecomment-1288857212>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AASTVHQUXZFZCYJQ2RTT523WEZTPXANCNFSM6AAAAAAQJQDJ7M>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Claudenw committed:
Merge remote-tracking branch 'upstream/Collections-385_BitCountProducer_operation_is_not_clearly_defined' into Collections-384_BitCountProducer_operation_is_not_clearly_defined

Claudenw commented:
> Last commit was on Oct 21st. Is this correct or do you need to push again?

Found the issue on my system.  Should be updated now.

aherbert commented on src/main/java/org/apache/commons/collections4/bloomfilter/BitCountProducer.java:
```diff
@@ -57,12 +88,22 @@ static BitCountProducer from(IndexProducer idx) {
             }
+
+            @Override
+            public int[] asIndexArray() {
```
I presume this change is required otherwise the test suite fails.

aherbert commented on src/main/java/org/apache/commons/collections4/bloomfilter/HasherCollection.java:
```diff
@@ -106,6 +120,23 @@ public boolean forEachIndex(IntPredicate consumer) {
+    public IndexProducer absoluteUniqueIndices(final Shape shape) {
+        return consumer -> {
+            Objects.requireNonNull(consumer, "consumer");
+            return uniqueIndices(shape).forEachIndex(IndexFilter.create(shape, consumer));
```
IIUC this can exceed the capacity of the IndexFilter. The uniqueIndices(shape) will output up to k indices for each hasher in the collection (eliminating duplicates). Thus if we have two hashers with shape(k=5, n=1000) we could output 10 indices from the combined unique indices. Then we create an IndexFilter for shape which could be optimised only to check up to k=5 indices.

This throws an index out of bounds exception from IndexFilter.ArrayTracker.

```Java
@Test
public void testAbsoluteUniqueIndices() {
    int[] actual = new HasherCollection(
        new IncrementingHasher(1, 1),
        new IncrementingHasher(10, 1)
    ).absoluteUniqueIndices(Shape.fromKM(5, 1000)).asIndexArray();
    int[] expected = IntStream.concat(
            IntStream.range(1, 1 + 5),
            IntStream.range(10, 10 + 5)
        ).toArray();
    Assertions.assertArrayEquals(expected, actual);
}
```

When I update HasherCollection absoluteUniqueIndices to use a fake shape that forces the IndexTracker.BitMapTracker implementation then the above tests passes:
```Java
public IndexProducer absoluteUniqueIndices(final Shape shape) {
    return consumer -> {
        Objects.requireNonNull(consumer, "consumer");
        return uniqueIndices(shape).forEachIndex(IndexFilter.create(
                Shape.fromKM(shape.getNumberOfBits(), shape.getNumberOfBits()), consumer));
    };
}
```

I think that this can be easily solved by having a new package-private factory constructor in IndexFilter to choose the implementation that can handle exceeding the number of hash functions in the shape:

```Java
static IntPredicate create(int numberOfBits, IntPredicate consumer) {
    // Choose the BitMapTracker implementation to handle any number of hash functions
```

Or you could create a new Shape as:
```Java
public IndexProducer absoluteUniqueIndices(final Shape shape) {
    return consumer -> {
        Objects.requireNonNull(consumer, "consumer");
        return uniqueIndices(shape).forEachIndex(IndexFilter.create(
                Shape.fromKM(shape.getNumberOfHashFunctions() * hashers.size(),
                             shape.getNumberOfBits()), consumer));
    };
}
```
Perhaps the later is simpler.

Claudenw replied:
implemented later version with a check for hashers.size() == 0

aherbert commented on src/main/java/org/apache/commons/collections4/bloomfilter/HasherCollection.java:
```diff
@@ -141,29 +172,16 @@ public boolean forEachIndex(IntPredicate consumer) {
-                offset += ary.length;
-            }
-            return result;
+            int[] result = new int[shape.getNumberOfHashFunctions()*hashers.size()];
```
This is a much better implementation.

Can you add whitespace around the multiply ` * `.

Claudenw replied:
fixed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/BitCountProducerFromAbsoluteUniqueHasherCollectionTest.java:
```diff
@@ -0,0 +1,50 @@
+    protected BitCountProducer createProducer() {
+        // hasher has collisions and wraps
+        return BitCountProducer.from(new HasherCollection(
+                new IncrementingHasher(1, 1),
```
This tests a saturation of the number of bits. We should really have a test for non saturation, no collisions just to make sure that works as well.

Claudenw replied:
Changed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/IndexProducerFromAbsoluteUniqueHasherCollectionTest2.java:
```diff
@@ -0,0 +1,44 @@
+ */
+package org.apache.commons.collections4.bloomfilter;
+
+public class IndexProducerFromAbsoluteUniqueHasherCollectionTest2 extends AbstractIndexProducerTest {
```
Remove the 2 from the name suffix

Claudenw replied:
fixed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/BitCountProducerFromAbsoluteUniqueHasherCollectionTest.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    protected int getBehaviour() {
+        // Hasher allows duplicates and may be unordered
+        return 0;
```
Should this have flags for distinct?

Claudenw replied:
Yes

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/BitCountProducerFromUniqueHasherCollectionTest.java:
```diff
@@ -0,0 +1,50 @@
+ */
+package org.apache.commons.collections4.bloomfilter;
+
+
```
Remove empty line

Claudenw replied:
fixed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/BitCountProducerFromUniqueHasherTest.java:
```diff
@@ -0,0 +1,42 @@
+    @Override
+    protected int getBehaviour() {
+        // Hasher allows duplicates and may be unordered
+        return AS_ARRAY_DISTINCT;
```
Update the comment as this does not allow duplicates

Q. Is forEach also distinct?

Claudenw replied:
yes, fixed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/BitCountProducerFromUniqueHasherCollectionTest.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    protected int getBehaviour() {
+        // Hasher allows duplicates and may be unordered
```
`// HasherCollection allows ...`

Claudenw replied:
fixed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/EnhancedDoubleHasherTest.java:
```diff
@@ -35,6 +36,12 @@ protected Hasher createEmptyHasher() {
         return NullHasher.INSTANCE;
     }
 
+
```
Remove extra line

Claudenw replied:
fixed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/IndexProducerFromAbsoluteUniqueHasherCollectionTest2.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    protected int getBehaviour() {
+        // HasherCollection allows duplicates and may be unordered
```
Update comment as duplicates are not allowed from absoluteUniqueIndices.

Q. Should forEach also be distinct?

Claudenw replied:
Updataed

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/UniqueIndexProducerFromHasherCollectionTest.java:
```diff
@@ -36,4 +36,11 @@ protected int getBehaviour() {
         return 0;
     }
+
+
```
Remove extra line

Claudenw replied:
Done

aherbert reviewed (changes requested):
This looks good. However there is a bug in the new HasherCollection.absoluteUniqueIndices method due to how the IndexFilter is created. I've suggested two quick fixes to get around this.

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/HasherCollection.java:
```diff
@@ -90,6 +90,20 @@ public IndexProducer indices(final Shape shape) {
+     * and returning concatenated IndexProducer from each.</p>
+     *
+     * <p>A BitCountProducer generated from this IndexProducer is equivalent to a BitCountProducer from a
+     * counting Bloom filter that was constructed from the contained hashers.<p>
```
fixed

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/HasherCollection.java:
```diff
@@ -90,6 +90,20 @@ public IndexProducer indices(final Shape shape) {
+     *
+     * <p>This method may return duplicates if the collection of unique values from each of the contained
+     * hashers contain duplicates.  This is equivalent to creating Bloom filters for each contained hasher
+     * and returning concatenated IndexProducer from each.</p>
```
fixed

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/HasherCollection.java:
```diff
@@ -90,6 +90,20 @@ public IndexProducer indices(final Shape shape) {
     }
 
+    /**
+     * Creates an IndexProducer of comprising the unique indices from each of the contained
```
done

Claudenw commented on src/main/java/org/apache/commons/collections4/bloomfilter/Hasher.java:
```diff
@@ -49,7 +49,8 @@ public interface Hasher {
      *
      * <p>This is like the `indices(Shape)` method except that it adds the guarantee that no
-     * duplicate values will be returned</p>
+     * duplicate values will be returned.  The indices produced are equivalent those returned
```
fixed

Claudenw commented:
In all cases where `AS_ARRAY_ORDERED` is set `FOR_EACH_ORDERED` is also set, and the `FOR_EACH_COUNT_ORDERED` is therefore set as well.

In all cases where `AS_ARRAY_DISTINCT` is set `FOR_EACH_DISTINCT` is also set, and the `FOR_EACH_COUNT_DISTINCT` is therefore set as well.

I propose collapsing this into 2 statics:
`INDICES_ORDERED` : replaces `AS_ARRAY_ORDERED`, `FOR_EACH_ORDERED`, and `FOR_EACH_COUNT_ORDERED`.
`INDICES_DISTINCT`: replaces `AS_ARRAY_DISTINCT`, `FOR_EACH_DISTINCT` and `FOR_EACH_COUNT_DISTINCT`.

By "replaces" I mean remove all instances of the replaced constants (not make them equivalent).  
Add documentation that states the testing assumes the ORDERED and DISTINCT properties carry across `asArray()`, `forEachIndex()` and `forEachCount()`.

Also, since the `AbstractBitCountProducerTest ` extends `AbstractIndexProducerTest` we can remove all `IndexProducerFrom_X_Test` implementations where there is a `BitCountProducerFrom_X_Test`

@aherbert unless you have an objection, I'll proceed with these changes.

Claudenw committed:
Updated as per requests

aherbert commented:
Hi @Claudenw, consolidating all the behaviour flags is a good idea. It will avoid future tests failing to set them correctly. I'll wait until you have done that for a review.

Note: I have reservations about the HasherCollection and the requirement for an `absoluteUniqueIndices` method. In summary:

1. I think it should be resolved after this PR
2. Since the HasherCollection violates the IndexProducer contract by outputting more than k indices, we may need to revise the contract and javadoc as being able to output up to a multiple of k indices for a Shape. This would be helped by a size-type method in the IndexProducer interface to state what count of k will be output.
3. The method `absoluteUniqueIndices` may be better served by a HasherSet, or some other naming variant, that ensures all indices output by the IndexProducer are unique.

Claudenw committed:
Removed duplicate tests and static properties

Claudenw committed:
removed accidental files

Claudenw commented:
@aherbert The changes are in.  Have a look please

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/DefaultIndexProducerTest.java:
```diff
@@ -47,8 +52,7 @@ public boolean forEachIndex(IntPredicate predicate) {
 
     @Override
     protected int getBehaviour() {
-        // The default method streams a BitSet so is distinct and ordered.
```
Note: This test has removed some testing of the default methods in IndexProducer.

It should be named `IndexProducerFromIndexArrayTest`. It is specifically targeting the default implementation of the IndexProducer when created from an index array. In that case the `asIndexArray()` method returns a clone of the original indices used to create it.

A `DefaultIndexProducerTest` would test the default implementation of `asIndexArray()` and look like this.

```Java
public class DefaultIndexProducerTest extends AbstractIndexProducerTest {

    /** Make forEachIndex unordered and contain duplicates. */
    private int[] values = {10, 1, 10, 1};

    @Override
    protected int[] getExpectedIndices() {
        return values;
    }

    @Override
    protected IndexProducer createProducer() {
        return new IndexProducer() {
            @Override
            public boolean forEachIndex(IntPredicate predicate) {
                Objects.requireNonNull(predicate);
                for (int i : values) {
                    if (!predicate.test(i)) {
                        return false;
                    }
                }
                return true;
            }
        };
    }

    @Override
    protected IndexProducer createEmptyProducer() {
        return new IndexProducer() {
            @Override
            public boolean forEachIndex(IntPredicate predicate) {
                Objects.requireNonNull(predicate);
                return true;
            }
        };
    }

    @Override
    protected int getBehaviour() {
        // The default method streams a BitSet so is distinct and ordered.
        // However the forEachIndex may not be distinct and ordered and
        // the test cannot distinguish the two cases.
        return 0;
    }
}
```

Note that this highlights a problem with consolidating all the behaviour flags. Currently the test suite cannot be used to test the asIndexArray behaviour is different from the forEachIndex behaviour.

I think we should revert the consolidation of the flags. However to avoid the duplication of flag setting through the various test cases we can change the AbstractIndexProducerTest test to do this:
```Java
public abstract class AbstractIndexProducerTest {

    protected static final int ORDERED = 0x1;
    protected static final int DISTINCT = 0x2;

    protected abstract int getAsIndexArrayBehaviour();

    protected int getForEachIndexBehaviour() {
        return getAsIndexArrayBehaviour();
    }

}
```
And AbtractBitCountProducerTest:
```Java
public abstract class AbstractBitCountProducerTest extends AbstractIndexProducerTest {

    protected int getForEachCountBehaviour() {
        return getAsIndexArrayBehaviour();
    }

}
```

Then get the correct flag in the relevant test. It provides the ability to override each test behaviour individually but will default to the same behaviour if not explicitly changed.

Doing this change (DefaultIndexProducerTest and IndexProducerFromIndexArrayTest) should remove the requirement for the additional test `testFromIndexArray`. The method `testFromBitMapProducer` is already covered in IndexProducerFromBitMapProducerTest.

Claudenw replied:
Corrected.  I also added `DefaultBitCountProducerTest` and removed duplicate test `UniqueIndexProducerFromHashCollectionTest` which was effectively a duplicate of `BitCountProducerFromUniqueHasherCollectionTest`

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/IndexProducerFromBitmapProducerTest.java:
```diff
@@ -49,10 +49,15 @@ protected IndexProducer createProducer() {
     protected int getBehaviour() {
         // Bit maps will be distinct. Conversion to indices should be ordered.
-        return FOR_EACH_DISTINCT | FOR_EACH_ORDERED | AS_ARRAY_DISTINCT | AS_ARRAY_ORDERED;
+        return INDICES_DISTINCT | INDICES_ORDERED | INDICES_DISTINCT | INDICES_ORDERED;
```
Duplicates

aherbert reviewed (changes requested):
Hi @Claudenw 

I think the consolidation of the ordered and distinct flags has reduced the flexibility of the test suite. It also led to you duplicating the flags in all test classes that override getBehaviour().

However the previous flexibility can be reinstated with my suggestion to have a method to get the behaviour for each of: asIndexArray; forEachIndex; forEachCount. By default this would return the same behaviour for all of them. But it can be configured to change and allow testing different behaviour for the methods such as that done in the default implementation of IndexProducer.

Claudenw committed:
Modifications as requested.

Added DefaultBitCountProducerTest
removed duplicate test UniqueIndexProducerFromHashCollectionTest was
effectivly a duplicate of BitCountProducerFromUniqueHasherCollectionTest

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/DefaultBitCountProducerTest.java:
```diff
@@ -0,0 +1,94 @@
+
+            @Override
+            public boolean forEachCount(BitCountConsumer consumer) {
+                int[] vals = values.clone();
```
There are no comments as to why you are doing this. Was it for testing?

I suggest leaving forEachCount to not be ordered and match the forEachIndex order. This should test the default implementations of BitCountProducer, not add in extra tests for convenience.

Claudenw replied:
I should have reverted that.  It was to test that the forEachCount tests used the proper behaviour.

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/DefaultBitCountProducerTest.java:
```diff
@@ -0,0 +1,94 @@
+    @Override
+    protected int getAsIndexArrayBehaviour() {
+        // The default method streams a BitSet so is distinct and ordered.
+        // However the forEachIndex may not be distinct and ordered and
```
This comment is outdated. The comment should be moved to `getForEachIndexBehaviour` as:
```Java
// The forEachIndex output is intentionally unordered and contains duplicates
```

If you wish to test forEachCount can use different flags then add some comments:
```Java
// Force to be ordered
int[] vals = values.clone();
Arrays.sort(vals);

// ...

// The forEachCount output is intentionally ordered and contains duplicates
```

aherbert commented on src/test/java/org/apache/commons/collections4/bloomfilter/DefaultIndexProducerTest.java:
```diff
@@ -36,22 +38,41 @@ protected int[] getExpectedIndices() {
-    protected int getBehaviour() {
+    protected int getAsIndexArrayBehaviour() {
+        // The default method streams a BitSet so is distinct and ordered.
+        // However the forEachIndex may not be distinct and ordered and
```
Again can you move this comment to `getForEachIndexBehaviour` as:
```Java
// The forEachIndex output is intentionally unordered and contains duplicates
```

aherbert reviewed (changes requested):
Hi @Claudenw 

Thanks for the changes. They look good. I requested a few comment changes to indicate what the tests are intentionally trying to test.

Claudenw committed:
Changes as per review.
